pkg/kf/commands/apps/push.go

```diff
@@ -351,3 +367,73 @@ func calculateScaleBounds(instances int, minScale, maxScale *int) (int, int, err
 	}
 
 }
+
+func createRoute(appName string, routeStr string, namespace string) (*v1alpha1.Route, error) {
+	hostname, domain, path, err := parseRouteStr(routeStr)
+	if err != nil {
+		return nil, err
+	}
+
+	r := &v1alpha1.Route{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "Route",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name: v1alpha1.GenerateName(
+				hostname,
+				domain,
+				path,
+			),
+		},
+		Spec: v1alpha1.RouteSpec{
+			Hostname:            hostname,
+			Domain:              domain,
+			Path:                path,
+			KnativeServiceNames: []string{appName},
+		},
+	}
+
+	return r, nil
+}
+
+// parseRouteStr parses a route URL into a hostname, domain, and path
+func parseRouteStr(routeStr string) (string, string, string, error) {
+	u, err := url.Parse(routeStr)
+	if err != nil {
+		return "", "", "", fmt.Errorf("failed to parse route: %s", err)
+	}
+	if u.Scheme == "" {
+		// Parsing URLs without schemes causes the hostname and domain to incorrectly be empty.
+		// We handle this by assuming the route has a HTTP scheme if scheme is not provided.
+		u, err = url.Parse("http://" + routeStr)
+		if err != nil {
+			return "", "", "", fmt.Errorf("failed to parse route: %s", err)
+		}
+	}
+
+	parts := strings.SplitN(u.Hostname(), ".", 3)
+
+	var hostname string
+	var domain string
+	var path string
+
+	if len(parts) == 3 {
+		// Has hostname
+		if parts[0] == "www" {
```

Comment:
should this be handled server side (e.g., webhook) instead of client side?

Reply:
Can we ignore this rule? I think devs should be able to shoot themselves in the foot if they want.

Reply:
Assuming users have routes including "www" in their manifest, then we have to handle the parsing of this somehow, otherwise the hostname and domain will just be silently incorrect. For now I left this logic on the client-side (and added moving the "www" over from the hostname to the domain name) and added a check in the validation webhook that confirms that the hostname isn't set to "www". In the future, we can move all of this over to the server side